aten/src/ATen/cudnn/Handle.cpp

```diff
@@ -17,8 +15,25 @@ struct Handle {
   Handle() : handle(NULL) {
     AT_CUDNN_CHECK(cudnnCreate(&handle));
   }
+  // std::vector.emplace() and push_back() may route through temporaries and call
+  // copy/move constructors along the way.  If this is the case, we don't want
+  // the destructors of temporaries to call cudnnDestroy on the handle.
+  // We can achieve safety (for the narrow case of stashing within STL containers)
+  // by defining copy and move constructors that transfer cudnnDestroy
+  // responsibility to the latest constructed object.  This is NOT a substitute for
+  // full-blown reference counting, but reference counting may be overkill here.
+  // Another alternative is to wrap the saved Handles in unique_ptrs, i.e.,
+  // unordered_map<int, vector<unique_ptr<Handle>>> created_handles;
+  void transfer(Handle& rhs) {
+    handle = rhs.handle;
+    rhs.handle = NULL;
+  }
+  Handle(Handle& rhs) { transfer(rhs); }
+  Handle(Handle&& rhs) { transfer(rhs); }
+  // operator= takes argument by value
+  Handle& operator=(Handle rhs) { transfer(rhs); return *this; }
```

Comment:
Careful. You have to check that rhs is not this; otherwise you'll leak the handle.

Reply:
rhs is taken by value, following one of my (favorite stackoverflow posts)[https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom].
So within the function body, `rhs` will never be `*this`.  You could write something like
```
Handle handle_lvalue;
handle_lvalue = std::move(handle_lvalue);
```
but I think in this case `handle_lvalue` will still end up owning the original handle it constructed (the `rhs` argument to `operator=` will be move-constructed from `handle_lvalue`, and steal the handle, then within `operator=`, `handle_lvalue` will steal it back.
https://godbolt.org/z/cN2NQL
^ I don't think godbolt will let me run interactively, but when I compile and run on my local machine, it prints 1.

Reply:
Oh, I didn't realize you were using the copy-and-swap idiom here. I think you're the first user of this in PyTorch :) Worth a brief comment, I suppose.